File: advent_of_code/2015/01/day01.hs
coerce :: Char -> Int
coerce '(' = 1
coerce ')' = -1
coerce _ = 0

part1 :: String -> Int
part1 = sum . fmap coerce

part2 :: String -> Int
part2 = length . takeWhile (>=0) . scanl (+) 0 . fmap coerce

main :: IO ()
main = do
  input <- readFile "input.txt"
  putStrLn $ "2015 Day 1, Part 1: " ++ show (part1 input)
  putStrLn $ "2015 Day 1, Part 2: " ++ show (part2 input)

File: scala/aoc/src/main/scala/aoc2015/Day01.scala
package aoc2015

import common.Day

object Day01 extends Aoc2015Base:
  override val id = "01"
  override type Input = List[Char]
  override type Output = Int

  override def parse(input: String): List[Char] =
    input.trim().toList

  override def part1(input: Input): Output =
    input.map:
      case '(' => 1
      case ')' => -1
    .sum

  override def part2(input: Input): Output =
    input.map:
      case '(' => 1
      case ')' => -1
    .scan(0)(_ + _)
    .takeWhile(_ >= 0)
    .length


@main
def main(): Unit = Day.run(Day01)

